Feature/58 accessibility (#253)
reactjs/it.reactjs.org: content/docs/accessibility.md

/>
```

## HTML semantico {#semantic-html}
L'HTML semantico costituisce la base per l'accessibilità in un'applicazione web. Usare diversi elementi HTML per rafforzare il significato dell'informazione nel vostro sito, spesso può portare ad avere accessibilità gratis. 

- [MDN HTML elements reference](https://developer.mozilla.org/en-US/docs/Web/HTML/Element)

Alcune volte spezziamo la semantica di HTML quando inseriamo elementi come `<div>` all'interno di JSX solo per fare in modo che il nostro codice funzioni, specialmente quando lavoriamo con le liste  (`<ol>`, `<ul>` e `<dl>`) e tabelle `<table>`.

> In questi casi possiamo usare i [Fragments di React](/docs/fragments.html) per raggruppare più elmenti insieme.

Reviewer: ```suggestion
In questi casi possiamo usare i [Fragments di React](/docs/fragments.html) per raggruppare più elementi insieme.
```

}
```

Per maggiori informazioni fai riferimento alla documentazione sui [Fragments di React](/docs/fragments.html).

## Form accessibili {#accessible-forms}

### Labeling {#labeling}
Ogni elemento di un form, come ad esempio `<input>` e `<textarea>`, necessitano di essere etichettati come accessibili. É necessario mettere label descrittive, dato che queste vengono processate dai lettori di schermi.

Le seguenti risorse mostrano come raggiungere tale scopo:

> - [W3C mostra come etichettare gli elementi](https://www.w3.org/WAI/tutorials/forms/labels/)

Author: ```suggestion
- [W3C mostra come etichettare elementi](https://www.w3.org/WAI/tutorials/forms/labels/)
```

- [WebAIM mostra come etichettare elementi](https://webaim.org/techniques/forms/controls)
- [La Paciello Group spiega come mettere nomi accessibili](https://www.paciellogroup.com/blog/2017/04/what-is-an-accessible-name/)

> Sebbene tutte queste pratiche standard HTML possano essere usate in React, nota che l'attributo `for` in JSX viene scritto come `htmlFor`:

Author: ```suggestion
Sebbene tutte queste pratiche standard possano essere utilizzate in React, tieni presente che l'attributo `for` in JSX viene scritto come `htmlFor`:
```

```javascript{1}
<label htmlFor="namedInput">Name:</label>
<input id="namedInput" type="text" name="name"/>
```

### Notifica degli errori all'utente {#notifying-the-user-of-errors}

> Gli errori devono essere compresi da tutti gli utenti. I link seguenti mostrano come esporre testi di errore ai lettori di schermo:

Author: ```suggestion
Gli errori devono essere compresi da tutti gli utenti. Le seguenti risorse mostrano come esporre testi di errore ai lettori di schermo:
```

- [La W3C mostra le notifiche utente](https://www.w3.org/WAI/tutorials/forms/notifications/)
- [La WebAIM guarda alla validazione dei form](https://webaim.org/techniques/formvalidation/)

## Focus Control {#focus-control}

> Assicurati che la tua applicazione web sia totalmente compatibile con le operazioni fatte da tastiera:

Author: ```suggestion
Assicurati che la tua applicazione web sia totalmente utilizzabile tramite tastiera:
```

- [La WebAIM parla di accessibilità tramite tastiera](https://webaim.org/techniques/keyboard/)

### Keyboard focus and focus outline {#keyboard-focus-and-focus-outline}

> Il focus della tastiera fa riferimento all'elemento del DOM selezionato per accettare un input da tastiera. É semplicemente riconoscibile come un contorno sul campo di input come quello che si vede nell'immagine seguente:

Author: ```suggestion
Il focus della tastiera fa riferimento all'elemento del DOM selezionato per accettare un input. É semplicemente riconoscibile come un contorno sul campo di input come quello che si vede nell'immagine seguente:
```

<img src="../images/docs/keyboard-focus.png" alt="Blue keyboard focus outline around a selected link." />

É consigliabile usare il CSS per rimuovere questo contorno, ad esempio con `outline: 0`, se e solo se verrà rimpiazzato con un contorno in uno stile diverso.

> ### Mechanisms to skip to desired content {#mechanisms-to-skip-to-desired-content}

Author: Manca traduzione

Fornisci all'utente un meccanismo per saltare da una sezione all'altra dell'applicazione in quanto questo aiuta a velocizzare le operazioni di navigazione tramite tastiera.

> Skiplinks or Skip Navigation Links are hidden navigation links that only become visible when keyboard users interact with the page. They are very easy to implement with internal page anchors and some styling:

Author: Manca traduzione

- [WebAIM - Skip Navigation Links](https://webaim.org/techniques/skipnav/)

> Utilizza elementi di riferimento e ruoli, come ad esempio `<main>` and `<aside>`, per creare delle regioni nella pagina per permettere all'utente di navigare da una sezione all'altra.

Author: ```suggestion
Utilizza elementi di riferimento e ruoli, come ad esempio `<main>` e `<aside>`, per creare delle regioni, all'interno della pagina, per permettere all'utente di navigare da una sezione all'altra.
```

Per avere maggiori informazioni sull'uso di questi elementi per aumentare l'accessibilità si consiglia questa lettura:

- [Punti di riferimento accessibili](https://www.scottohara.me/blog/2018/03/03/landmarks.html)

> ### Programmatically managing focus {#programmatically-managing-focus}

Author: Manca traduzione

> Un'applicazione React modifica continuamente il DOM dell'applicazione durante la sua esecuzione, quindi qualche volte è possibile che il focus della tastiera venga perso oppure sia su un elemento diverso da quello che ci si aspettava. Per correggere questo comportamento errato è necessario intervenire programmaticamente. Ad esempio resettando il focus della tastiera sul bottone che ha aperto una finestra modale dopo che è stata chiusa.

Author: ```suggestion
Un'applicazione React modifica continuamente il DOM durante la sua esecuzione, quindi qualche volte è possibile che il focus della tastiera venga perso oppure si trovi su un elemento diverso da quello che ci si aspettava. Per correggere questo comportamento errato è necessario intervenire programmaticamente, ad esempio resettando il focus della tastiera sul bottone che ha aperto una finestra modale dopo che è stata chiusa.
```

}
```

Per maggiori informazioni fai riferimento alla documentazione sui [Fragments di React](/docs/fragments.html).

## Form accessibili {#accessible-forms}

### Labeling {#labeling}

> Ogni elemento di un form, come ad esempio `<input>` e `<textarea>`, necessitano di essere etichettati come accessibili. É necessario mettere label descrittive, dato che queste vengono processate dai lettori di schermi.

Reviewer: Il soggetto, elemento, e' singolare
```suggestion
Ogni elemento di un form, come ad esempio `<input>` e `<textarea>`, necessita di essere etichettato come accessibile. É necessario mettere label descrittive, dato che queste vengono processate dai lettori di schermi.
```

 }
 ```

> Qualche volta un componente padre necessita di fare focus su un elemento in un componente figlio. E' possibile raggiungere questo scopo [esponendo il refs del DOM al componente padre](/docs/refs-and-the-dom.html#exposing-dom-refs-to-parent-components) attraverso delle speciali prop, nel componente figlio, che inoltra il ref ai nodi figli.

Author: ```suggestion
Qualche volta un componente padre necessita di fare focus su un elemento di un componente figlio. E' possibile raggiungere questo scopo [esponendo il refs del DOM al componente padre](/docs/refs-and-the-dom.html#exposing-dom-refs-to-parent-components) attraverso delle speciali prop, nel componente figlio, che inoltra il ref ai nodi figli.
```

Author: Chiedo consiglio per tradurre meglio l'ultimo pezzo di frase

this.inputElement.current.focus();
```

Quando usiamo degli HOC per estendere il comportamento dei componenti, è raccomandabile di [inoltrare i ref](/docs/forwarding-refs.html) al componente "wrappato" usando la funzione di React `forwardRef`. Se un componente, di tipo HOC, di terze parti non implementata l'inoltro del ref, il pattern spiegato precedentemente può essere usato come seconda alternativa.

Un buon esempio di gestione del focus è [react-aria-modal](https://github.com/davidtheclark/react-aria-modal). Questo è un esempio di piena accessibilità di una finestra modale. Non solo mette il focus iniziale sul bottone cancel
(prevenendo un'attivazione accidentale dell'azione di successo) ma "imprigiona" il focus all'interno della finestra modale e lo rimette sull'elemento che inizialmente ha causato l'apertura della modale.

>Nota:
>

> >Sebbene questa sia una caratteristica di accessibilità molto importante, è anche una tecnica che dovrebbe essere usata con giudizio. Usalo per aggiustare il focus della tastiera quando è disturbato, non cercare di 

Author: ```suggestion
>Sebbene questa sia una caratteristica di accessibilità molto importante, è anche una tecnica che dovrebbe essere usata con giudizio. Usalo per aggiustare il focus della tastiera quando è rotto piutoosto che cercare di anticipare le mosse dell'utente 
```

> >anticipare le mosse dell'utente.

Author: ```suggestion
```

> ## Mouse and pointer events {#mouse-and-pointer-events}

Author: Manca traduzione

> Assicurati che tutte le funzionalità accessibili da mouse siano anche accessibili utilizzando solamente la tastiera. Dipendere solamente da elementi di puntamento portano gli utenti che utilizzano solo la tastiera a non utilizzare la tua applicazione.

Author: ```suggestion
Assicurati che tutte le funzionalità accessibili da mouse siano anche accessibili utilizzando solamente la tastiera. Dipendere solamente da elementi di puntamento porta gli utenti che utilizzano solo la tastiera a non utilizzare la tua applicazione.
```

this.inputElement.current.focus();
```

Quando usiamo degli HOC per estendere il comportamento dei componenti, è raccomandabile di [inoltrare i ref](/docs/forwarding-refs.html) al componente "wrappato" usando la funzione di React `forwardRef`. Se un componente, di tipo HOC, di terze parti non implementata l'inoltro del ref, il pattern spiegato precedentemente può essere usato come seconda alternativa.

Un buon esempio di gestione del focus è [react-aria-modal](https://github.com/davidtheclark/react-aria-modal). Questo è un esempio di piena accessibilità di una finestra modale. Non solo mette il focus iniziale sul bottone cancel
(prevenendo un'attivazione accidentale dell'azione di successo) ma "imprigiona" il focus all'interno della finestra modale e lo rimette sull'elemento che inizialmente ha causato l'apertura della modale.

>Nota:
>

> >Sebbene questa sia una caratteristica di accessibilità molto importante, è anche una tecnica che dovrebbe essere usata con giudizio. Usalo per aggiustare il focus della tastiera quando è rotto piuttosto che cercare di anticipare le mosse dell'utente 

Reviewer: Il soggetto, tecnica, e' femminile
```suggestion
>Sebbene questa sia una caratteristica di accessibilità molto importante, è anche una tecnica che dovrebbe essere usata con giudizio. Usala per aggiustare il focus della tastiera quando è rotto piuttosto che cercare di anticipare le mosse dell'utente 
```

}
```

> Questo può funzionare bene per gli utenti che usano strumenti di puntamento, come ad esempio un mouse, ma utilizzando solamente la tastiera porta ad una rottura della funzionalità quando ci spostiamo sul successivo elemento col pulsante di Tab, in quanto l'oggetto `window` non riceverà mai l'evento di `click`. Questo errato comportamento può portare a nascondere una certa funzionalità dell'applicazione e di conseguenza di ha un allontamento degli utente.

Author: ```suggestion
Questo può funzionare bene per gli utenti che usano strumenti di puntamento, come ad esempio un mouse, ma utilizzando solamente la tastiera porta ad una rottura della funzionalità quando ci spostiamo sul successivo elemento col pulsante di Tab, in quanto l'oggetto `window` non riceverà mai l'evento di `click`. Questo errato comportamento può portare a nascondere una certa funzionalità dell'applicazione e di conseguenza ad un allontamento degli utenti.
```

<img src="../images/docs/outerclick-with-keyboard.gif" alt="A toggle button opening a popover list implemented with the click outside pattern and operated with the keyboard showing the popover not being closed on blur and it obscuring other screen elements." />

> La stessa funzionalità può essere ottenuta semplicemente utilizzando in modo appropriato eventi come `onBlur` e `onFocus`: 

Author: ```suggestion
La stessa funzionalità può essere ottenuta semplicemente utilizzando in modo appropriato eventi come `onBlur` e `onFocus`:
```

}
```

> Il codice appena visto mostra come esporre la funzionalità sia per utenti che usano strumenti di puntamento sia per utenti che usano la tastiera. Nota anche che è stata aggiunta la props `aria-*` per fornire supporto ai lettori di schermo degli utenti. Per semplicità l'evento della tastiera per abilitare l'interazione con il popover tramite `tasti freccia` non è stato implementato.

Author: ```suggestion
Il codice appena visto mostra come esporre la funzionalità sia per utenti che usano strumenti di puntamento sia per utenti che usano la tastiera. Nota anche che è stata aggiunta la props `aria-*` per fornire supporto agli utenti che utilizzano i lettori di schermo. Per semplicità l'evento della tastiera per abilitare l'interazione con il popover tramite `tasti freccia` non è stato implementato.
```

<img src="../images/docs/blur-popover-close.gif" alt="A popover list correctly closing for both mouse and keyboard users." />

> Questo è solo un esempio di molti dove dipendere solamente da dispositivi di puntamento e eventi del mouse portano ad una rottura delle funzionalità delle tastiera degli utenti. Testare sempre con la tastiera evidenzia immediatamente eventuali problemi che dovranno essere corretti utilizzando eventi della tastiera.

Author: ```suggestion
Questo è solo un esempio di molti casi dove dipendere solamente da dispositivi di puntamento e eventi del mouse portano ad una rottura delle funzionalità per gli utenti che utilizzano la tastiera. Testare sempre con la tastiera evidenzia immediatamente eventuali problemi che dovranno essere corretti utilizzando eventi della tastiera.
```

> ## More Complex Widgets {#more-complex-widgets}

Author: Manca traduzione

Una più complessa esperienza utente non significa avere meno accessibilità. L'accessibilità si ottiene più facilmente scrivendo HTML, ma anche i widget più complessi possono essere resi accessibili.

E' richiesta una conoscenza di [ARIA Roles](https://www.w3.org/TR/wai-aria/#roles) così come [ARIA States and Properties](https://www.w3.org/TR/wai-aria/#states_and_properties).
Questi sono strumenti con attributi HTML che sono pienamente supportati in JSX e che ti permettono di costruire componenti React completamente accessibili.

Ogni tipo di widget ha uno specifico pattern e gli utenti si aspettano che funzioni in un determinato modo;

- [WAI-ARIA Authoring Practices - Design Patterns and Widgets](https://www.w3.org/TR/wai-aria-practices/#aria_ex)
- [Heydon Pickering - ARIA Examples](https://heydonworks.com/article/practical-aria-examples/)
- [Inclusive Components](https://inclusive-components.design/)

> ## Other Points for Consideration {#other-points-for-consideration}

Author: Manca traduzione

### Impostazioni della lingua {#setting-the-language}

> Scrivete in un linguaggio naturale i testi delle pagine in modo tale che i lettori di schermo possono effettuare le corrette impostazioni della voce:

Author: Chiedo aiuta per una migliore traduzione

- [WebAIM - Document Language](https://webaim.org/techniques/screenreader/#language)

### Impostazioni del titolo del documento {#setting-the-document-title}

Imposta il `<title>` del documento in modo che descriva in modo corretto il contenuto della pagina in quanto questo garantisce che l'utente rimanga sempre consapevole del contenuto della pagina in cui si trova:

> - [WCAG - Understanding the Document Title Requirement](https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-title.html)

Author: Manca traduzione

Per farlo in React basta semplicemente usare il [componente Document Title](https://github.com/gaearon/react-document-title).

> ### Color contrast {#color-contrast}

Author: Manca traduzione

Assicuratevi che tutti i testi leggibili sul vostro sito abbiamo colori adatti in modo tale da rimanere sempre visibili anche ad utenti con carenze visive:

> - [WCAG - Understanding the Color Contrast Requirement](https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html)

Author: Manca traduzione

> - [Everything About Color Contrast And Why You Should Rethink It](https://www.smashingmagazine.com/2014/10/color-contrast-tips-and-tools-for-accessibility/)

Author: Manca traduzione

> - [A11yProject - What is Color Contrast](https://a11yproject.com/posts/what-is-color-contrast/)

Author: Manca traduzione

Può risultare molto noioso calcolare tutte le combinazioni di colori per tutti i casi, per questo motivo è possibile [calcolare l'intera palette di colori accessibili con Colorable](https://jxnblk.com/colorable/).

Entrambi gli strumenti menzionati prima (aXe e WAVE) includono test sul contrasto dei colori.

> Se vuoi estenedere per conto tuo il test sul contrasto dei colori puoi utilizzare i seguenti strumenti:

Author: ```suggestion
Se vuoi estendere per conto tuo il test sul contrasto dei colori puoi utilizzare i seguenti strumenti:
```

- [WebAIM - Color Contrast Checker](https://webaim.org/resources/contrastchecker/)
- [The Paciello Group - Color Contrast Analyzer](https://www.paciellogroup.com/resources/contrastanalyser/)

## Strumenti di sviluppo e test {#development-and-testing-tools}

Ci sono numerosi strumenti che possono essere utilizzati e che ti "assistono" nella creazione di applicazioni web accessibili.

> ### The keyboard {#the-keyboard}

Author: Manca traduzione

Il test più facile da fare, e anche quello più importante, è di navigare tutta l'applicazione utilizzando esclusivamente la tastiera. Fai questi passi:

1. Scollega il mouse.
1. Usa i tasti di `Tab` e `Shift+Tab` per la navigazione.
1. Utilizza il tasto di `Enter` per interagire con gli elementi.
1. Quando richiesto utilizza i tasti freccia per interagire con elementi come ad esempio i menù a tendina.

> ### Development assistance {#development-assistance}

Author: Manca traduzione

  }
  ```

### Testare l'accessibilità all'interno del browser {#testing-accessibility-in-the-browser}

Esistono numerosi tool che eseguono verifiche di accessibilità sulla pagine web nel tuo browser. Utilizza uno di questo in combinazione con altri strumenti qua menzionati, in quanto questi testano solamente l'accessibilità "tecnica" del tuo codice HTML.

#### aXe, aXe-core e react-axe {#axe-axe-core-and-react-axe}

> Deque Systems offre [aXe-core](https://github.com/dequelabs/axe-core) per test end-to-end e automatici di accessibilità. Questo modulo include alcune integrazioni per Selenium.

Reviewer: Meglio?
```suggestion
Deque Systems offre [aXe-core](https://github.com/dequelabs/axe-core) per test di accessibilità end-to-end automatizzati. Questo modulo include alcune integrazioni per Selenium.
```

> ### Testing accessibility in the browser {#testing-accessibility-in-the-browser}

Author: Manca traduzione

Esistono numerosi tool che eseguono verifiche di accessibilità sulla pagine web nel tuo browser. Utilizza uno di questo in combinazione con altri strumenti qua menzionati, in quanto questi testano solamente l'accessibilità "tecnica" del tuo codice HTML.

#### aXe, aXe-core and react-axe {#axe-axe-core-and-react-axe}

Deque Systems offre [aXe-core](https://github.com/dequelabs/axe-core) per test end-to-end e automatici di accessibilità. Questo modulo include alcune integrazioni per Selenium.

[L'Accessibility Engine](https://www.deque.com/products/axe/) (abbreviato con aXe), è un'estensione per il proprio browser costruita con `aXe-core`.

Puoi anche usare il modulo [react-axe](https://github.com/dylanb/react-axe) per vedere errori e problemi vari di accessibilità direttamente nella console in fase di sviluppo e debug.

#### WebAIM WAVE {#webaim-wave}

Il [Web Accessibility Evaluation Tool](https://wave.webaim.org/extension/) è un'altra estensione per il browser riguardante l'accessibilità.

> #### Accessibility inspectors and the Accessibility Tree {#accessibility-inspectors-and-the-accessibility-tree}

Author: Manca traduzione
